pkg/apiserver/registry/inventory/virtualmachine/rest.go

```diff
@@ -0,0 +1,177 @@
+	"k8s.io/apimachinery/pkg/apis/meta/internalversion"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/apiserver/pkg/endpoints/request"
+	"k8s.io/apiserver/pkg/registry/rest"
+
+	runtimev1alpha1 "antrea.io/nephe/apis/runtime/v1alpha1"
+	"antrea.io/nephe/pkg/controllers/config"
+	"antrea.io/nephe/pkg/controllers/inventory"
+	"antrea.io/nephe/pkg/controllers/inventory/common"
+	"antrea.io/nephe/pkg/controllers/inventory/store"
+)
+
+// REST implements rest.Storage for VirtualMachine Inventory.
+type REST struct {
+	cloudInventory inventory.Interface
+	logger         logger.Logger
+}
+
+var (
+	_ rest.Scoper  = &REST{}
+	_ rest.Getter  = &REST{}
+	_ rest.Watcher = &REST{}
+	_ rest.Lister  = &REST{}
+)
+
+// NewREST returns a REST object that will work against API services.
+func NewREST(cloudInventory inventory.Interface, l logger.Logger) *REST {
+	return &REST{
+		cloudInventory: cloudInventory,
+		logger:         l,
+	}
+}
+
+func (r *REST) New() runtime.Object {
+	return &runtimev1alpha1.VirtualMachine{}
+}
+
+func (r *REST) NewList() runtime.Object {
+	return &runtimev1alpha1.VirtualMachine{}
+}
+
+func (r *REST) ShortNames() []string {
+	return []string{"vm"}
+}
+
+func (r *REST) Get(ctx context.Context, name string, _ *metav1.GetOptions) (runtime.Object, error) {
+	ns, ok := request.NamespaceFrom(ctx)
+	if !ok || len(ns) == 0 {
+		return nil, errors.NewBadRequest("Namespace cannot be empty.")
+	}
+
+	namespacedName := ns + "/" + name
+	vm, ok := r.cloudInventory.GetVmByKey(namespacedName)
+	if !ok {
+		return nil, errors.NewNotFound(runtimev1alpha1.Resource("virtualmachine"), name)
+	}
+	return vm, nil
+}
+
+func (r *REST) List(ctx context.Context, options *internalversion.ListOptions) (runtime.Object, error) {
+	// List only supports two types of input options
+	// 1. All namespaces
+	// 2. Labelselector with only the specific namespace, the only valid labelselectors are "cpa.name=<accountname>"
+	//    and "cpa.namespace=<accountNamespace>"
+	// 3. Specific Namespace
+	accountName := ""
+	accountNamespace := ""
+	if options != nil && options.LabelSelector != nil && options.LabelSelector.String() != "" {
+		labelSelectorStrings := strings.Split(options.LabelSelector.String(), ",")
+		for _, labelSelectorString := range labelSelectorStrings {
+			labelKeyAndValue := strings.Split(labelSelectorString, "=")
+			if labelKeyAndValue[0] == config.LabelCloudAccountName {
+				accountName = labelKeyAndValue[1]
+			} else if labelKeyAndValue[0] == config.LabelCloudAccountNamespace {
+				accountNamespace = labelKeyAndValue[1]
+			} else {
+				return nil, errors.NewBadRequest("unsupported label selector, only region label selector is supported")
+			}
+		}
+	}
+
+	var objs []interface{}
+	namespace, _ := request.NamespaceFrom(ctx)
+	if namespace == "" {
+		objs = r.cloudInventory.GetAllVms()
+	} else if accountName != "" {
```

Comment:
with new change, it should be cpa-name + cpa-namespace.. both label should be specified..

Reply:
Done